nbbuild/misc/prepare-bundles/src/main/java/org/netbeans/prepare/bundles/PrepareBundles.java

```diff
@@ -0,0 +1,295 @@
+            for (Path module : ds) {
+                if (".bin".equals(module.getFileName().toString())) continue;
+                if ("@types".equals(module.getFileName().toString())) continue;
+                Path packageJson = module.resolve("package.json");
+
+                if (!Files.isReadable(packageJson)) {
+                    throw new IllegalStateException("Cannot find package.json for: " + module.getFileName());
+                }
+
+                String packageJsonText = readString(packageJson);
+                Map<String, Object> packageJsonData = new Gson().fromJson(packageJsonText, HashMap.class);
+                String name = (String) packageJsonData.get("name");
+                String version = (String) packageJsonData.get("version");
+                String description = (String) packageJsonData.get("description");
+                String homepage = (String) packageJsonData.get("homepage");
+                String licenseKey = (String) packageJsonData.get("license");
+
+                String licenseText = null;
+
+                for (String l : LICENSE_FILE_NAMES) {
+                    if (Files.isReadable(module.resolve(l))) {
+                        licenseText = readString(module.resolve(l));
+                        break;
+                    }
+                }
+
+                if (licenseText == null) {
+                    String hardcodedLicenseName = name + "-" + version + "-license";
+                    URL hardcodedLicense = PrepareBundles.class.getResource(hardcodedLicenseName);
+                    if (hardcodedLicense == null ){
+                        throw new IllegalStateException("Cannot find license for: " + module.getFileName());
+                    }
+                    StringBuilder licenseTextBuffer = new StringBuilder();
+                    try (InputStream in = hardcodedLicense.openStream();
+                         Reader r = new InputStreamReader(in, StandardCharsets.UTF_8)) {
+                        int read;
+                        while ((read = r.read()) != (-1)) {
+                            licenseTextBuffer.append((char) read);
+                        }
+                    }
+                    licenseText = licenseTextBuffer.toString();
+                }
+                
+                Path thirdpartynoticestxt = module.resolve("thirdpartynotices.txt");
+
+                if (Files.isReadable(thirdpartynoticestxt)) {
+                    licenseText = "Parts of this work are licensed:\n" +
+                                  licenseText +
+                                  "\n\n" +
+                                  "Parts of this work are licensed:\n" +
+                                  readString(thirdpartynoticestxt);
+                } else {
+                    licenseText = licenseText;
+                }
+
+                Path noticestxt = module.resolve("CopyrightNotice.txt");
```

Comment:
This seems to be problematic. One sample I found from a quick look at the generated NOTICE file was this:

```
/*! *****************************************************************************
Copyright (c) Microsoft Corporation. All rights reserved. 
Licensed under the Apache License, Version 2.0 (the "License"); you may not use
this file except in compliance with the License. You may obtain a copy of the
License at http://www.apache.org/licenses/LICENSE-2.0  
 
THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE, 
MERCHANTABLITY OR NON-INFRINGEMENT. 
 
See the Apache Version 2.0 License for specific language governing permissions
and limitations under the License.
***************************************************************************** */
```

The author did not understand the purpose of the NOTICE file and if we'd integrate this file by hand, I'd request it to be completely removed, as nothing of the text has a place in the NOTICE file.

I admit though, that it is difficult to detect and override. Maybe notices should only done by hand?

Reply:
I thought this should be there for the "Copyright (c) Microsoft Corporation. All rights reserved." line. But I could either drop this (if not needed at all), or use a hardcoded set of notices if needed (e.g. with just the single line or so.)

Reply:
I read this as the normal license header for the application of licenses - you state the Copyright, the years you touched the file in, the copyright holder and the license. Nothing more - but it was only one of the entries, others were more suble but also held more information than necessary. I have not yet a good idea how to fix this...